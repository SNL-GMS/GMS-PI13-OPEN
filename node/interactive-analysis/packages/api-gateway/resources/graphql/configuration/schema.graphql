
# Interface for analyst configuration used to determine fields based on roles
type AnalystConfiguration {
  logLevel: LogLevel!
  defaultNetwork: String!
  sohStationGroupNames: [SOHStationGroupNameWithPriority]!
  defaultFilters: [WaveformFilter]!
  redisplayPeriod: Int!
  reprocessingPeriod: Int!
  acknowledgementQuietDuration: Int!
  availableQuietDurations: [Int]!
  sohStationStaleTimeMS: Int!
  sohHistoricalDurations: [Int]!
  systemMessageLimit: Int!
}

# SOH StationGroup and Priority interface definition
type SOHStationGroupNameWithPriority {
  name: String
  priority: Int
}

# Represents a Waveform Filter
type WaveformFilter {
  id: String!
  name: String!
  description: String!
  filterType: String!
  filterPassBandType: String!
  lowFrequencyHz: Float
  highFrequencyHz: Float
  order: Float!
  filterSource: String!
  filterCausality: String!
  zeroPhase: String!
  sampleRate: Float!
  sampleRateTolerance: Float!
  validForSampleRate: String!
  aCoefficients: [Float]
  bCoefficients: [Float]
  groupDelaySecs: Float
}

# The following event related queries are supported.
extend type Query {
    uiAnalystConfiguration: AnalystConfiguration
}
