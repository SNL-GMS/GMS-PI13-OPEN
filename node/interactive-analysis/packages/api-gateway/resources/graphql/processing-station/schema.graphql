# Represents a group of stations used for monitoring.
# This is the processing equivalent of the ReferenceNetwork.
type ProcessingStationGroup {
    name: String
    description: String
    stations: [ProcessingStation]
}

# Represents an installation of monitoring sensors for the purposes of processing.
# Multiple sensors can be installed at the same station.
type ProcessingStation {
    name: String
    type: StationType
    description: String
    location: Location
    channelGroups: [ProcessingChannelGroup]
    channels: [ProcessingChannel]
}

# ChannelGroupType enum represents the different groupings of channels
enum ChannelGroupType {
    PROCESSING_GROUP
    SITE_GROUP
}

# Represents a physical installation (e.g., building, underground vault, borehole)
# containing a collection of Instruments that produce Raw Channel waveform data.
type ProcessingChannelGroup {
    name: String
    description: String
    location: Location
    type: ChannelGroupType
    channels: [ProcessingChannel]
}

# Represents a source for unprocessed (raw) or processed (derived) time series data
# from a seismic, hydroacoustic, or infrasonic sensor.
type ProcessingChannel {
    name: String
    displayName: String
    canonicalName: String
    description: String
    station: String
    channelDataType: ChannelDataType
    nominalSampleRateHz: Int
    location: Location
    orientationAngles: Orientation
}

# Enumeration representing the different types of processing channels.
enum ChannelDataType {
  SEISMIC
  HYDROACOUSTIC
  INFRASOUND
  WEATHER
  DIAGNOSTIC_SOH
  DIAGNOSTIC_WEATHER
}

# Represents the orientation angles used in processing channels
type Orientation {
    horizontalAngleDeg: Float
    verticalAngleDeg: Float
}

#####################################################
# Resolver entries Mutation, Query and Subscriptions
# The following channel-related queries are supported.
extend type Query {

    # Retrieve the default station list for use in automated and interactive analysis
    defaultProcessingStations: [ProcessingStation]
}


