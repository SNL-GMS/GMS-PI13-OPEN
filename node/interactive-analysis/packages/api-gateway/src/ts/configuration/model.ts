import Immutable from 'immutable';
import { WaveformFilterDefinition } from '../waveform-filter/model';

/**
 * Interface for analyst configuration used to determine fields based on roles
 */
export interface AnalystConfiguration {
  readonly defaultNetwork: string;
  readonly defaultFilters: WaveformFilterDefinition[];
}

/**
 * Soh specific configuration
 */
export interface SohConfiguration {
  stationSohControlConfiguration: {
    reprocessingPeriod: string;
    displayedStationGroups: string[];
    rollupStationSohTimeTolerance: string;
  };
  stationSohMonitoringDisplayParameters: {
    redisplayPeriod: string;
    acknowledgementQuietDuration: string;
    availableQuietDurations: string[];
    sohStationStaleDuration: string;
    sohHistoricalDurations: string[];
  };
}

/**
 * Common configuration
 */
export interface CommonConfiguration {
  systemMessageLimit: number;
}

/**
 * Overall configuration
 */
export interface Configuration {
  /**
   * AnalystConfiguration keyed on user role
   */
  analystConfiguration: Immutable.Map<string, AnalystConfiguration>;

  /**
   * SohConfigurationOld
   */
  sohConfiguration: SohConfiguration;

  /**
   * CommonConfiguration keyed on user role
   */
  commonConfiguration: Immutable.Map<string, CommonConfiguration>;
}

/**
 * Selector interface for config service
 */
export interface Selector {
  criterion: string;
  value: string;
}

/**
 * Analyst configurations loaded from service
 */
export enum AnalystConfigs {
  DEFAULT = 'ui.analyst-settings'
}

/**
 * SOH configurations loaded from service
 */
export const SohConfig = 'ui.soh-settings';

/**
 * Common configurations loaded from service
 */
export enum CommonConfigs {
  DEFAULT = 'ui.common-settings'
}

/**
 * Default user role
 */
export const defaultUserRole = 'DEFAULT';
