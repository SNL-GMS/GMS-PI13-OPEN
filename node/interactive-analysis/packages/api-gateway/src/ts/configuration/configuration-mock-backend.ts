import { readJsonData } from '@gms/common-util';
import config from 'config';
import Immutable from 'immutable';
import path from 'path';
import { gatewayLogger as logger } from '../log/gateway-logger';
import { HttpMockWrapper } from '../util/http-wrapper';
import { resolveTestDataPaths } from '../util/test-data-util';
import {
  AnalystConfigs,
  AnalystConfiguration,
  CommonConfigs,
  CommonConfiguration,
  Selector,
  SohConfig,
  SohConfiguration
} from './model';

/**
 * Mock backend HTTP services providing access to configuration data. If mock services are enabled in the
 * configuration file, this module loads a test data set specified in the configuration file and configures
 * mock HTTP interfaces for the API gateway backend service calls.
 */

/** Declare a backend data store for the mock configuration backend */
let dataStore: Immutable.Map<string, any> = Immutable.Map<string, any>();

/**
 * Configure mock HTTP interfaces for a simulated set of configuration backend services.
 * @param httpMockWrapper The HTTP mock wrapper used to configure mock backend service interfaces
 */
export function initialize(httpMockWrapper: HttpMockWrapper) {
  logger.info('Initializing mock backend for analysis configuration data');

  if (!httpMockWrapper) {
    throw new Error(
      'Cannot initialize mock configuration services with undefined HTTP mock wrapper'
    );
  }

  // Load test data from the configured data set
  loadTestData();

  // Load the configuration backend service config settings
  const backendConfig = config.get('configuration.backend');
  httpMockWrapper.onMock(
    backendConfig.services.getAnalystConfiguration.requestConfig.url,
    getConfiguration
  );
  httpMockWrapper.onMock(
    backendConfig.services.getSohConfiguration.requestConfig.url,
    getSohConfiguration
  );
}

/**
 * Load test data into the mock backend data store from the configured test data set.
 * Put them into the data store map
 */
function loadTestData(): void {
  const dataPath = resolveTestDataPaths().additionalDataHome;
  const additionalTestData = config.get('testData.additionalTestData');

  // ui config
  const analystConfigurationFile = path.join(
    dataPath,
    additionalTestData.analystConfigurationFilename
  );
  logger.info(`Loading configuration test data from path: ${analystConfigurationFile}`);

  let defaultAnalystConfiguration: AnalystConfiguration;
  try {
    defaultAnalystConfiguration = readJsonData(analystConfigurationFile) as any;
    logger.info(`Mock backend configuration loaded.`);
  } catch (e) {
    logger.error(
      `Failed to read configuration data from files: ` +
        `${additionalTestData.analystConfigurationFilename}`
    );
  }
  dataStore = dataStore.set(AnalystConfigs.DEFAULT, defaultAnalystConfiguration);

  // common config
  const commonConfigurationFile = path.join(
    dataPath,
    additionalTestData.commonConfigurationFilename
  );
  logger.info(`Loading configuration test data from path: ${commonConfigurationFile}`);

  let commonConfiguration: CommonConfiguration;
  try {
    commonConfiguration = readJsonData(commonConfigurationFile) as any;
    logger.info(`Mock backend common configuration loaded.`);
  } catch (e) {
    logger.error(
      `Failed to read configuration data from files: ` +
        `${additionalTestData.commonConfigurationFilename}`
    );
  }
  dataStore = dataStore.set(CommonConfigs.DEFAULT, commonConfiguration);

  // soh ui config
  const uiSohConfigurationFile = path.join(dataPath, additionalTestData.uiSohConfigurationFileName);
  logger.info(`Loading soh configuration test data from path: ${uiSohConfigurationFile}`);
  let uiSohConfig: SohConfiguration;
  try {
    uiSohConfig = readJsonData(uiSohConfigurationFile) as any;
    logger.info(`Mock backend soh configuration loaded.`);
  } catch (e) {
    logger.error(
      `Failed to read configuration data from files: ` +
        `${additionalTestData.analystConfigurationFilename}`
    );
  }
  dataStore = dataStore.set(SohConfig, uiSohConfig);
}

/**
 * Retrieves the analyst configuration based on a user role
 * @param configName configuration to retrieve
 * @returns a configuration
 */
function getConfiguration(parameters: { configName: string; selectors: Selector[] }): any {
  // Handle undefined input
  if (!parameters.configName) {
    throw new Error('Unable to retrieve Configuration for undefined config name');
  }

  // Handle uninitialized data store
  handleUninitializedDataStore();
  const configuration = dataStore.get(parameters.configName);
  if (parameters.selectors.length <= 0) {
    return configuration;
  }
  // if there are selectors - we need to get processing config
  const monitorType = parameters.selectors[0].value;
  const monitorTypeConfiguration = configuration[monitorType];
  let selectorIndex = parameters.selectors.length - 1;
  let configToReturn = monitorTypeConfiguration[parameters.selectors[selectorIndex].value];
  while (selectorIndex > 1 && configToReturn === undefined) {
    selectorIndex--;
    configToReturn = monitorTypeConfiguration[parameters.selectors[selectorIndex].value];
  }
  // if no channel or station was found - get the default
  return configToReturn ? configToReturn : configuration[parameters.selectors[0].value].default;
}

/**
 * Retrieves the analyst configuration based on a user role
 * @param string of no consequence
 * @returns a configuration
 */
function getSohConfiguration(whoCares: string): any {
  // Handle uninitialized data store
  handleUninitializedDataStore();
  return dataStore.get(SohConfig);
}

/**
 * Handle cases where the data store has not been initialized.
 */
function handleUninitializedDataStore() {
  // If the data store is uninitialized, throw an error
  if (!dataStore) {
    throw new Error('Mock backend configuration data store has not been initialized');
  }
}
