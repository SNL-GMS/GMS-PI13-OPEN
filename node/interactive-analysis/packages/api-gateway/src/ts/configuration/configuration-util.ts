import { SohTypes } from '@gms/common-graphql';

/**
 * Maps monitor type to value type
 * @param monitorType monitor type to map to value type
 */
export const getValueTypeForMonitor = (
  monitorType: SohTypes.SohMonitorType
): SohTypes.SohValueType =>
  monitorType === SohTypes.SohMonitorType.LAG
    ? SohTypes.SohValueType.DURATION
    : SohTypes.SohValueType.PERCENT;
