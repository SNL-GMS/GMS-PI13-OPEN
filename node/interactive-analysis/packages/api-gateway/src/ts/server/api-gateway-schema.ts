import { makeExecutableSchema } from 'apollo-server-express';
import merge from 'lodash/merge';
import path from 'path';
import { resolvers as cacheResolvers } from '../cache/resolvers';
import { schema as cacheSchema } from '../cache/schema';
import { resolvers as channelSegmentResolvers } from '../channel-segment/resolvers';
import { schema as channelSegmentSchema } from '../channel-segment/schema';
import {
  extendedResolvers as commonExtendedResolvers,
  resolvers as commonResolvers
} from '../common/resolvers';
import { extendedSchema as commonExtendedSchema, schema as commonSchema } from '../common/schema';
import { resolvers as configurationResolvers } from '../configuration/resolvers';
import { schema as configurationSchema } from '../configuration/schema';
import { resolvers as dataAcquisitionStatusResolvers } from '../data-acquisition/resolvers';
import { schema as dataAcquisitionStatusSchema } from '../data-acquisition/schema';
import { schema as eventSchema } from '../event/model-and-schema/schema';
import { resolvers as eventResolvers } from '../event/resolvers';
import { resolvers as fkResolvers } from '../fk/resolvers';
import { schema as fkSchema } from '../fk/schema';
import { resolvers as qcMaskResolvers } from '../qc-mask/resolvers';
import { schema as qcMaskSchema } from '../qc-mask/schema';
import { resolvers as signalDetectionResolvers } from '../signal-detection/resolvers';
import { schema as signalDetectionSchema } from '../signal-detection/schema';
import { resolvers as sohStatusResolvers } from '../soh/resolvers';
import { schema as sohStatusSchema } from '../soh/schema';
import { resolvers as processingStationResolvers } from '../station/processing-station/resolvers';
import { schema as processingStationSchema } from '../station/processing-station/schema';
import { resolvers as referenceStationResolvers } from '../station/reference-station/resolvers';
import { schema as referenceStationSchema } from '../station/reference-station/schema';
import { resolvers as systemMessageResolvers } from '../system-message/resolvers';
import { schema as systemMessageSchema } from '../system-message/schema';
import { resolvers as userProfileResolvers } from '../user-profile/resolvers';
import { schema as userProfileSchema } from '../user-profile/schema';
import { resolvers as waveformResolvers } from '../waveform/resolvers';
import { schema as waveformSchema } from '../waveform/schema';
import { resolvers as workflowResolvers } from '../workflow/resolvers';
import { schema as workflowSchema } from '../workflow/schema';

/**
 * Global GraphQL schema definition for the entire API gateway
 */

// using the gateway logger
import { gatewayLogger as logger } from '../log/gateway-logger';
const objectPath = path.relative(process.cwd(), __filename);

// GraphQL schema definitions
logger.info('Creating graphql schema...', { module: objectPath });

const typeDefs = [
  commonSchema,
  commonExtendedSchema,
  systemMessageSchema,
  workflowSchema,
  processingStationSchema,
  referenceStationSchema,
  waveformSchema,
  signalDetectionSchema,
  eventSchema,
  qcMaskSchema,
  userProfileSchema,
  channelSegmentSchema,
  fkSchema,
  dataAcquisitionStatusSchema,
  sohStatusSchema,
  cacheSchema,
  configurationSchema
];

// Merge GraphQL resolvers from the schema domains
const resolvers = merge(
  commonResolvers,
  commonExtendedResolvers,
  systemMessageResolvers,
  workflowResolvers,
  processingStationResolvers,
  referenceStationResolvers,
  waveformResolvers,
  signalDetectionResolvers,
  eventResolvers,
  userProfileResolvers,
  qcMaskResolvers,
  channelSegmentResolvers,
  fkResolvers,
  dataAcquisitionStatusResolvers,
  sohStatusResolvers,
  cacheResolvers,
  configurationResolvers
);

// Build the GraphQL schema
export const schema = makeExecutableSchema({
  typeDefs,
  resolvers
});
