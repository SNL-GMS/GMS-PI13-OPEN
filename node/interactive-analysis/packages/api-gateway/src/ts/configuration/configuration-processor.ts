import { SOHStationGroupNameWithPriority } from '@gms/common-graphql/lib/graphql/ui-configuration/types';
import { getDurationMilliTime, sleep } from '@gms/common-util';
import { ApolloError } from 'apollo-server-core';
import config from 'config';
import Immutable from 'immutable';
import { gatewayLogger, gatewayLogger as logger } from '../log/gateway-logger';
import { HttpClientWrapper, HttpResponse } from '../util/http-wrapper';
import { WaveformFilterDefinition } from '../waveform-filter/model';
import * as configurationMockBackend from './configuration-mock-backend';
import {
  AnalystConfigs,
  AnalystConfiguration,
  CommonConfigs,
  CommonConfiguration,
  Configuration,
  defaultUserRole,
  Selector,
  SohConfiguration
} from './model';

/**
 * Singleton class used to retrieve Configuration attributes such as network and waveform filters
 * At start up a default network is fetched from the configuration service, if a user who logs in isn't the default
 * that users configuration will be fetched, at start up the station processors need a network and use a the default
 * network therefore the async method getDefaultNetwork is used, other wise after log in, can guarantee that
 * a users configuration will be loaded and the other get methods don't need to be awaited on.
 */
export class ConfigurationProcessor {
  /**
   * Returns the singleton instance of the cache processor.
   * @returns the instance of the cache processor
   */
  public static Instance(): ConfigurationProcessor {
    if (!ConfigurationProcessor.isInitialized) {
      ConfigurationProcessor.instance = new ConfigurationProcessor();
      ConfigurationProcessor.instance.initialize();
      ConfigurationProcessor.isInitialized = true;
    }
    return ConfigurationProcessor.instance;
  }

  /** The singleton instance */
  private static instance: ConfigurationProcessor;

  private static isInitialized: boolean = false;

  /**
   * The configuration
   */
  private readonly configuration: Configuration;

  /** Local configuration settings for reference stations */
  private readonly settings: any;

  /** HTTP client wrapper for communicating with backend services */
  private readonly httpWrapper: HttpClientWrapper;

  /**
   * Constructor - initialize the processor, loading settings and initializing the HTTP client wrapper.
   */
  private constructor() {
    // Load configuration settings
    this.settings = config.get('configuration');

    // Initialize an http client
    this.httpWrapper = new HttpClientWrapper();

    // Initialize the configuration
    this.configuration = {
      analystConfiguration: Immutable.Map(),
      sohConfiguration: undefined,
      commonConfiguration: Immutable.Map()
    };
  }

  /**
   * Gets the analyst configuration based on the role
   * @param role of the user from the user context
   * @returns a analyst configuration
   */
  public getAnalystConfiguration(role: string = defaultUserRole): AnalystConfiguration {
    if (this.configuration.analystConfiguration.has(role)) {
      return this.configuration.analystConfiguration.get(role);
    }
    throw new ApolloError(
      `Configuration error, could not find analyst configuration for role: ${role}`
    );
  }

  /**
   * Gets the common configuration based on the role
   * @param role of the user from the user context
   * @returns a common configuration
   */
  public getCommonConfiguration(role: string = defaultUserRole): CommonConfiguration {
    if (this.configuration.commonConfiguration.has(role)) {
      return this.configuration.commonConfiguration.get(role);
    }
    throw new ApolloError(
      `Configuration error, could not find common configuration for role: ${role}`
    );
  }

  /**
   * Gets the network based on the role
   * @param role of the user from the user context
   * @returns a network as a Promise<string>
   */
  public getNetworkForUser(role: string = defaultUserRole): string {
    if (this.configuration.analystConfiguration.has(role)) {
      return this.configuration.analystConfiguration.get(role).defaultNetwork;
    }
    throw new ApolloError(
      `Configuration error, could not find configuration default network for role: ${role}`
    );
  }

  /**
   * Gets the waveform filters based on the role
   * @param role of the user from the user context
   * @returns a waveform filters as Promise<WaveformFilterDefinition[]>
   */
  public getWaveformFiltersForUser(role: string = defaultUserRole): WaveformFilterDefinition[] {
    if (this.configuration.analystConfiguration.has(role)) {
      return this.configuration.analystConfiguration.get(role).defaultFilters;
    }
    throw new ApolloError(
      `Configuration error, could not find configuration default filters for role: ${role}`
    );
  }

  /**
   * Gets the soh configuration based on the role
   * @returns a soh configuration
   */
  public getSohConfiguration(): SohConfiguration {
    if (this.configuration.sohConfiguration) {
      return this.configuration.sohConfiguration;
    }
    throw new ApolloError(`Configuration error, could not find soh configuration`);
  }

  /**
   * Gets the soh configuration based on the role
   * @returns a soh configuration
   */
  public getStationGroupNamesWithPriorities(): SOHStationGroupNameWithPriority[] {
    if (!this.configuration.sohConfiguration) {
      return [];
    }
    return this.configuration.sohConfiguration.stationSohControlConfiguration.displayedStationGroups.map(
      (sg, count) => ({ name: sg, priority: count })
    );
  }

  /**
   * Gets the SOH Update Interval based on the role
   * @param role of the user from the user context
   * @returns the interval in seconds
   */
  public getSohUpdateIntervalForUser(role: string = defaultUserRole): number {
    if (this.configuration.sohConfiguration) {
      const sohParams = this.configuration.sohConfiguration.stationSohMonitoringDisplayParameters;
      return getDurationMilliTime(sohParams.redisplayPeriod);
    }
    throw new ApolloError(
      `Configuration error, could not find configuration redisplayPeriod for role: ${role}`
    );
  }

  /**
   * Gets the SOH quiet interval based on the role
   * @param role of the user from the user context
   * @returns the interval in seconds
   */
  public getSohDefaultQuietInterval(role: string = defaultUserRole): number {
    if (this.configuration.sohConfiguration) {
      const sohParams = this.configuration.sohConfiguration.stationSohMonitoringDisplayParameters;
      return getDurationMilliTime(sohParams.acknowledgementQuietDuration);
    }
    throw new ApolloError(
      `Configuration error, could not find configuration acknowledgementQuietDuration for role: ${role}`
    );
  }

  /**
   * Load and fetch the initial configurations for the given user role
   * @param role of the user from the user context
   */
  public async fetchConfiguration(): Promise<Configuration> {
    // if needed, fetch configurations
    await this.fetchAnalystConfiguration();
    await this.fetchSohConfiguration();
    await this.fetchCommonConfiguration();
    return this.configuration;
  }

  /**
   * Fetches AnalystConfiguration from the configuration service for the given unique configuration id
   * @param configuration the configuration id to retrieve
   */
  private async fetchAnalystConfiguration(
    role: string = defaultUserRole
  ): Promise<AnalystConfiguration> {
    if (this.configuration.analystConfiguration.has(role)) {
      return this.configuration.analystConfiguration.get(role);
    }

    // get the configuration id based on the user role
    const configuration: string = AnalystConfigs[role];

    logger.info(
      `Fetching analyst configuration data: role: ${role} configuration: ${configuration}`
    );
    const analystConfiguration = await this.internalFetchSubConfiguration<AnalystConfiguration>(
      configuration
    );
    this.configuration.analystConfiguration = this.configuration.analystConfiguration.set(
      role,
      analystConfiguration
    );
    return this.configuration.analystConfiguration.get(role);
  }

  /**
   * Fetches CommonConfiguration from the configuration service for the given unique configuration id
   * @param configuration the configuration id to retrieve
   */
  private async fetchCommonConfiguration(
    role: string = defaultUserRole
  ): Promise<CommonConfiguration> {
    if (this.configuration.commonConfiguration.has(role)) {
      return this.configuration.commonConfiguration.get(role);
    }

    // get the configuration id based on the user role
    const configuration: string = CommonConfigs[role];

    logger.info(
      `Fetching common configuration data: role: ${role} configuration: ${configuration}`
    );
    const commonConfiguration = await this.internalFetchSubConfiguration<CommonConfiguration>(
      configuration
    );
    this.configuration.commonConfiguration = this.configuration.commonConfiguration.set(
      role,
      commonConfiguration
    );
    return this.configuration.commonConfiguration.get(role);
  }

  /**
   * Fetches SohConfiguration from the configuration service for the given unique configuration id
   * @param configuration the configuration to retrieve
   */
  private async fetchSohConfiguration(): Promise<SohConfiguration> {
    // add data to the post because our service needs something in the body
    const postBodyData = 'test';
    // Retrieve the request configuration for the service call
    const requestConfiguration = this.settings.backend.services.getSohConfiguration.requestConfig;
    // Call the service and process the response data and add configuration to map
    const sohConfiguration = await this.callFetchSohConfiguration(
      requestConfiguration,
      postBodyData
    );
    this.configuration.sohConfiguration = sohConfiguration;
    return this.configuration.sohConfiguration;
  }

  /**
   * calls the webservice to get the sohconfiguration
   * @param requestConfiguration configuration given to the http request wrapper
   * @param postBodyData string to add to the boyd of the request,
   * the call needs a string in the body of the request
   */
  private readonly callFetchSohConfiguration = async (
    requestConfiguration: any,
    postBodyData: string
  ): Promise<SohConfiguration> =>
    this.httpWrapper
      .request<SohConfiguration>(requestConfiguration, postBodyData)
      .then((response: HttpResponse<SohConfiguration>) => response.data)
      .catch(async e => {
        logger.warn(`Error failed to connect to soh configuration service: ${e}`);
        const timeoutMS = 2000;
        return sleep<Promise<SohConfiguration>>(timeoutMS).then(async () => {
          logger.info(`Retrying to connect to configuration service`);
          logger.info(`retry config:: ${JSON.stringify(requestConfiguration)}`);
          return this.callFetchSohConfiguration(requestConfiguration, postBodyData);
        });
      })

  /**
   * Hits configuration service to get back an AnalystConfiguration based on role
   * @param configuration configuration to retrieve
   */
  private async internalFetchSubConfiguration<T>(
    configuration: string,
    selectors: Selector[] = []
  ): Promise<T> {
    // Build the query to be encoded as query string parameters
    const query = {
      configName: configuration,
      selectors
    };

    // Retrieve the request configuration for the service call
    const requestConfig = this.settings.backend.services.getAnalystConfiguration.requestConfig;

    logger.debug(
      `Requesting Analyst configuration query ${JSON.stringify(query)} config ${JSON.stringify(
        requestConfig
      )}`
    );

    // Call the service and process the response data and add configuration to map
    return this.httpWrapper
      .request<T>(requestConfig, query)
      .then((response: HttpResponse<T>) => response.data)
      .catch(async e => {
        logger.warn(`Error failed to connect to configuration service: ${e}`);
        const timeoutMS = 2000;

        return sleep<Promise<AnalystConfiguration>>(timeoutMS).then(async () => {
          logger.info(`Retrying to connect to configuration service`);
          return this.internalFetchSubConfiguration<T>(configuration, selectors);
        });
      });
  }

  /**
   * Initialize the configuration processor, fetching configuration from the backend.
   * This function sets up a mock backend if configured to do so.
   */
  private initialize() {
    try {
      logger.info(
        'Initializing the configuration processor - Mock Enable: %s',
        this.settings.backend.mock.enable
      );

      // If service mocking is enabled, initialize the mock backend
      if (this.settings.backend.mock.enable) {
        configurationMockBackend.initialize(this.httpWrapper.createHttpMockWrapper());
      }
    } catch (error) {
      gatewayLogger.error(`Failed to initialize the Configuration processor: ${error}`);
      throw error;
    }
  }
}
