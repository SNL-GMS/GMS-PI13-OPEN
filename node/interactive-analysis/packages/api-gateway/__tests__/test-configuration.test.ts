import { graphql } from 'graphql';
import { ConfigurationProcessor } from '../src/ts/configuration/configuration-processor';
import { schema } from '../src/ts/server/api-gateway-schema';
import { ProcessingStationProcessor } from '../src/ts/station/processing-station/processing-station-processor';
import { userContext } from './__data__/user-profile-data';

beforeAll(async () => setupTest());

/**
 * Sets up test by loading SDs
 */
async function setupTest() {
  // Simulates an initial login request to fetch configuration
  await ConfigurationProcessor.Instance().fetchConfiguration();
  await ProcessingStationProcessor.Instance().fetchStationData();
}

describe('User Configuration', () => {
  // tslint:disable-next-line: max-line-length
  it('A users gets their default network and default filters loaded in the Configuration Processor after login', async () => {
    const query = `
    query {
      uiAnalystConfiguration{
        logLevel
        defaultNetwork
        redisplayPeriod
        reprocessingPeriod
        acknowledgementQuietDuration
        availableQuietDurations
        sohStationStaleTimeMS
        sohHistoricalDurations
        sohStationGroupNames {
          name
          priority
        }
        defaultFilters {
          id
          name
          description
          filterType
          filterPassBandType
          lowFrequencyHz
          highFrequencyHz
          order
          filterSource
          filterCausality
          zeroPhase
          sampleRate
          sampleRateTolerance
          groupDelaySecs
          validForSampleRate
        }
      }
    }
    `;

    // Execute the GraphQL query
    const rootValue = {};
    const result = await graphql(schema, query, rootValue, userContext);
    const { data } = result;
    // Compare response to snapshot
    expect(data).toMatchSnapshot();
  });
});
