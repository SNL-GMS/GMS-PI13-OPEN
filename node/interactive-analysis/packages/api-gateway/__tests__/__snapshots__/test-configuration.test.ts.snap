// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Configuration A users gets their default network and default filters loaded in the Configuration Processor after login 1`] = `
Object {
  "uiAnalystConfiguration": Object {
    "acknowledgementQuietDuration": 300000,
    "availableQuietDurations": Array [
      300000,
      900000,
      3600000,
      86400000,
      604800000,
    ],
    "defaultFilters": Array [
      Object {
        "description": "Hamming FIR Filter Band Pass, 0.70-2.00 Hz",
        "filterCausality": "CAUSAL",
        "filterPassBandType": "BAND_PASS",
        "filterSource": "SYSTEM",
        "filterType": "FIR_HAMMING",
        "groupDelaySecs": 1.2,
        "highFrequencyHz": 2,
        "id": "48fd578e-e428-43ff-9f9e-62598e7e6ce6",
        "lowFrequencyHz": 0.7,
        "name": "HAM FIR BP 0.70-2.00 Hz",
        "order": 48,
        "sampleRate": 20,
        "sampleRateTolerance": 0.05,
        "validForSampleRate": "true",
        "zeroPhase": "false",
      },
      Object {
        "description": "Hamming FIR Filter Band Pass, 0.70-2.00 Hz",
        "filterCausality": "CAUSAL",
        "filterPassBandType": "BAND_PASS",
        "filterSource": "SYSTEM",
        "filterType": "FIR_HAMMING",
        "groupDelaySecs": 0.6,
        "highFrequencyHz": 2,
        "id": "0351d87a-fde9-43e2-8754-84a1d797fbc1",
        "lowFrequencyHz": 0.7,
        "name": "HAM FIR BP 0.70-2.00 Hz",
        "order": 48,
        "sampleRate": 40,
        "sampleRateTolerance": 0.05,
        "validForSampleRate": "true",
        "zeroPhase": "false",
      },
      Object {
        "description": "Hamming FIR Filter Band Pass, 1.00-3.00 Hz",
        "filterCausality": "CAUSAL",
        "filterPassBandType": "BAND_PASS",
        "filterSource": "SYSTEM",
        "filterType": "FIR_HAMMING",
        "groupDelaySecs": 1.2,
        "highFrequencyHz": 3,
        "id": "56893621-d9a1-4cfd-830a-13969f9d2aad",
        "lowFrequencyHz": 1,
        "name": "HAM FIR BP 1.00-3.00 Hz",
        "order": 48,
        "sampleRate": 20,
        "sampleRateTolerance": 0.05,
        "validForSampleRate": "true",
        "zeroPhase": "false",
      },
      Object {
        "description": "Hamming FIR Filter Band Pass, 1.00-3.00 Hz",
        "filterCausality": "CAUSAL",
        "filterPassBandType": "BAND_PASS",
        "filterSource": "SYSTEM",
        "filterType": "FIR_HAMMING",
        "groupDelaySecs": 0.6,
        "highFrequencyHz": 3,
        "id": "8bab76bd-db1a-4226-a151-786577adb3d8",
        "lowFrequencyHz": 1,
        "name": "HAM FIR BP 1.00-3.00 Hz",
        "order": 48,
        "sampleRate": 40,
        "sampleRateTolerance": 0.05,
        "validForSampleRate": "true",
        "zeroPhase": "false",
      },
      Object {
        "description": "Hamming FIR Filter Band Pass, 4.00-8.00 Hz",
        "filterCausality": "CAUSAL",
        "filterPassBandType": "BAND_PASS",
        "filterSource": "SYSTEM",
        "filterType": "FIR_HAMMING",
        "groupDelaySecs": 1.19999999999,
        "highFrequencyHz": 8,
        "id": "db5e61a0-b8dc-48a7-b56c-5e497925e89c",
        "lowFrequencyHz": 4,
        "name": "HAM FIR BP 4.00-8.00 Hz",
        "order": 48,
        "sampleRate": 20,
        "sampleRateTolerance": 0.05,
        "validForSampleRate": "true",
        "zeroPhase": "false",
      },
      Object {
        "description": "Hamming FIR Filter Band Pass, 4.00-8.00 Hz",
        "filterCausality": "CAUSAL",
        "filterPassBandType": "BAND_PASS",
        "filterSource": "SYSTEM",
        "filterType": "FIR_HAMMING",
        "groupDelaySecs": 0.6,
        "highFrequencyHz": 8,
        "id": "b92d6ade-3e69-42b6-82e5-f9290f50120f",
        "lowFrequencyHz": 4,
        "name": "HAM FIR BP 4.00-8.00 Hz",
        "order": 48,
        "sampleRate": 40,
        "sampleRateTolerance": 0.05,
        "validForSampleRate": "true",
        "zeroPhase": "false",
      },
    ],
    "defaultNetwork": "demo",
    "logLevel": "error",
    "redisplayPeriod": 5000,
    "reprocessingPeriod": 20,
    "sohHistoricalDurations": Array [
      21600000,
      43200000,
      86400000,
    ],
    "sohStationGroupNames": Array [
      Object {
        "name": "All_2",
        "priority": 0,
      },
      Object {
        "name": "All_1",
        "priority": 1,
      },
      Object {
        "name": "EurAsia",
        "priority": 2,
      },
      Object {
        "name": "Wrapped",
        "priority": 3,
      },
      Object {
        "name": "I_To_Z",
        "priority": 4,
      },
      Object {
        "name": "CD1.1",
        "priority": 5,
      },
      Object {
        "name": "OthCont",
        "priority": 6,
      },
      Object {
        "name": "A_To_H",
        "priority": 7,
      },
    ],
    "sohStationStaleTimeMS": 300000,
  },
}
`;
